platform/security/filter/security-filter-csrf/src/main/java/org/codice/ddf/security/filter/csrf/CsrfFilter.java

```diff
@@ -100,44 +122,101 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
-        respondForbidden(
-            httpResponse,
-            "Cross-site check failure: Request did not have required X-Requested-With header.");
-        return;
-      }
+    // Execute CSRF check if user is accessing /services
+    if (targetContextPath.startsWith(SERVICE_CONTEXT)
+        && doSystemProtectionFilter(
+            httpResponse, targetContextPath, requestMethod, userAgentHeader)) {
+      return;
     }
 
     // All checks passed
     chain.doFilter(request, response);
   }
 
+  private boolean doBrowserProtectionFilter(
+      HttpServletRequest httpRequest,
+      HttpServletResponse httpResponse,
+      String targetContextPath,
+      String requestMethod) {
+    String sourceOrigin = httpRequest.getHeader(ORIGIN_HEADER);
+    String sourceReferer = httpRequest.getHeader(REFERER_HEADER);
+    String csrfHeader = httpRequest.getHeader(CSRF_HEADER);
+
+    // Reject if no origin or referer header is present on the request
+    if (sourceOrigin == null && sourceReferer == null) {
+      respondForbidden(
+          httpResponse,
+          "Cross-site check failure: Incoming request did not have an Origin or Referer header.");
+      return true;
+    }
+
+    // Reject if neither the Origin nor Referer header are a trusted authority
+    if (!isTrustedAuthority(sourceOrigin) && !isTrustedAuthority(sourceReferer)) {
+      respondForbidden(
+          httpResponse,
+          "Cross-site check failure: Neither the Origin nor Referer header matched a system internal or external name.");
+      return true;
+    }
+
+    // Check for presence of anti-CSRF header if the path is not whitelisted
+    String whitelistKey = getMultivaluedMapKey(BROWSER_PROTECTION_WHITELIST, targetContextPath);
+
+    if (csrfHeader == null
+        && !(whitelistKey != null
+            && BROWSER_PROTECTION_WHITELIST.get(whitelistKey).contains(requestMethod))) {
+
+      respondForbidden(
+          httpResponse,
+          "Cross-site check failure: Request did not have required X-Requested-With header.");
+      return true;
+    }
+    return false;
+  }
+
+  private boolean doSystemProtectionFilter(
+      HttpServletResponse httpResponse,
+      String targetContextPath,
+      String requestMethod,
+      String userAgentHeader) {
+
+    String whitelistKey = getMultivaluedMapKey(SYSTEM_PROTECTION_WHITELIST, targetContextPath);
+    String whitelistUserAgent =
+        USER_AGENT_WHITELIST
+            .stream()
+            .filter(regex -> Pattern.matches(regex, userAgentHeader))
+            .findFirst()
+            .orElse(null);
+
+    if (userAgentHeader != null
+        && whitelistUserAgent == null
+        && !(whitelistKey != null
+            && SYSTEM_PROTECTION_WHITELIST.get(whitelistKey).contains(requestMethod))) {
+      respondForbidden(
+          httpResponse,
+          "Cross-site check failure: Request did not have required X-Requested-With header.");
```

Comment:
✏️ Sorry, I should've caught this on the first pass, but this message doesn't really apply here. It should probably say something about it not being system-to-system communication or having an unauthorized user agent header or something.